specs/core/1_shard-data-chains.md

```diff
@@ -17,595 +17,380 @@
-    )
+#### Block header
 
-    # Process and update block data fees
-    add_fee(state, beacon_state, proposer_index, state.basefee * len(block.core.data) // SHARD_BLOCK_SIZE_LIMIT)
-    QUOTIENT = SHARD_BLOCK_SIZE_LIMIT * BASEFEE_ADJUSTMENT_FACTOR
-    if len(block.core.data) > SHARD_BLOCK_SIZE_TARGET:
-        state.basefee += Gwei(max(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
-    elif len(block.core.data) < SHARD_BLOCK_SIZE_TARGET:
-        state.basefee -= Gwei(max(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
-    state.basefee = Gwei(max(
-        1,
-        min(
-            EFFECTIVE_BALANCE_INCREMENT // EPOCHS_PER_SHARD_PERIOD // SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH,
-            state.basefee,
-        )
-    ))
-
-    # Check total bytes
-    state.total_bytes += len(block.core.data)
-    assert block.core.total_bytes == state.total_bytes
-
-    # Update in-state block header
-    state.most_recent_block_core = ShardBlockCore(
-        slot=block.core.slot,
-        beacon_chain_root=block.core.beacon_chain_root,
-        parent_root=block.core.parent_root,
-        data_root=hash_tree_root(block.core.data),
-        state_root=Hash(),
-        total_bytes=block.core.total_bytes,
-        attester_bitfield=block.core.attester_bitfield,
+```python
+def process_shard_block_header(beacon_state: BeaconState, shard_state: ShardState, block: ShardBlock) -> None:
+    # Verify the shard number
+    assert block.shard == shard_state.shard
+    # Verify the slot number
+    assert block.slot == shard_state.slot
+    # Verify the beacon chain root
+    parent_epoch = compute_epoch_of_shard_slot(shard_state.latest_block_header.slot)
+    assert block.beacon_block_root == get_block_root(beacon_state, parent_epoch)
+    # Verify the parent root
+    assert block.parent_root == hash_tree_root(shard_state.latest_block_header)
+    # Save current block as the new latest block
+    shard_state.latest_block_header = ShardBlockHeader(
+        shard=block.shard,
+        slot=block.slot,
+        beacon_block_root=block.beacon_block_root,
+        parent_root=block.parent_root,
+        # `state_root` is zeroed and overwritten in the next `process_shard_slot` call
+        body_root=hash_tree_root(block.body),
+        block_size_sum=block.block_size_sum,
+        aggregation_bits=block.aggregation_bits,
+        attestations=block.attestations,
+        # `signature` is zeroed
     )
-
-    # Check state root
-    if validate_state_root:
-        assert block.core.state_root == hash_tree_root(state)
+    # Verify the sum of the block sizes since genesis
+    shard_state.block_size_sum += SHARD_HEADER_SIZE + len(block.body)
+    assert block.block_size_sum == shard_state.block_size_sum
+    # Verify proposer signature
+    proposer_index = get_shard_proposer_index(beacon_state, shard_state.shard, block.slot)
```

Comment:
Proposer is not guaranteed to be not slashed. In beacon blocks, we check for if slashed and prevent from proposing. Maybe we should filter in `get_shard_proposer_index` by active _and_ not slashed.

Also, maybe we should consider doing the same filtering in `get_beacon_proposer_index`

Reply:
With hindsight filtering for slashing in `get_beacon_proposer_index` makes more sense :)

Reply:
I've added the restriction that the proposer must not be slashed. We need to be very explicit about what `BeaconState` is when calling `shard_state_transition` because `BeaconState` changes on a slot-by-slot basis but we want the `BeaconState` at the last epoch boundary.